### DISCUSSION TITLE
compatibility issue. file stored with version 3.21 cannot be properly read

### DISCUSSION THREAD

**Discussion by rsm73S** [CGNS/CGNS#354] [CGNStalk]
I have an elder cga file which I able to open and properly read with an elder version of libcgns.a. Unfortunately I own only a binary file libcgns.a and the corresponding include file. Till now I have not been able to detect and check out the corresponding source into my working copy.
The cga file has version 3.21 and 64 bit precision.
I am getting errors while trying to read the same file with the current version (4.20 in cgnslib.h? not sure). The problem appears in cg_poly_elements_partial_read. The 7th argument, an array of type cgsize_t, contains now (with the current source code) the elements like these: 5, 0, 3, 0, 1, 0, 4 etc. All zeroes are wrong. If I watch the elder version, which works properly, the correct array elements must be like these: 5, 3, 1, 4 etc. It seems that the ADF somehow sets the wrong number of bytes / offsets while reading the corresponding section. Any idea why?

**Comment by mennodeij** [comment]
It looks like the 4.20 version you're using is compiled for cgsize_t of 32-bits, and the array contains 64-bits values. So each second value has zeroes when reading it as 32-bit values. The cgnslib.h header has a constant `CG_BUILD_64BIT`. If the value is 0, the library is 32-bits. 
As you're using a 64-bit file, you need to compile with `-DCGNS_ENABLE_64BIT=ON` on the cmake command prompt.

**Comment by rsm73S** [comment]
Thanks. But this is not true. I checked cgnslib,h. It includes in turn cgnstypes.h. There I see
#define CG_BUILD_64BIT  1
So my library must be already 64 bit.
The point is, something has been changed between now and an elder version (I assume even 4.10 because this stays in the header file of the library which works properly)

**Comment by mennodeij** [comment]
Ok, can you share the file maybe?

**Comment by rsm73S** [comment]
Yes, I attached an archive with the cga file and the woriking (old) version of libcgns.a
[archive.zip](https://github.com/CGNS/CGNS/files/9298062/archive.zip)

**Comment by rsm73S** [comment]
Here is also the new (not working) version of libcgns.a attached
[libcgns_new.zip](https://github.com/CGNS/CGNS/files/9298079/libcgns_new.zip)

**Comment by mennodeij** [comment]
I have just loaded the cga file into the cgnsview utility of a 64-bit CGNS build and I see no zeroes. The libcgns-new.a I cannot see anything about it.

**Comment by rsm73S** [comment]
Thanks. Either my library is compiled wrong or there is a special problem with cg_poly_elements_partial_read function or I use this function in the wrong way.
I have 2 questions
Is it possible to find out / check which precision does my library use (64 or 32 bit) at run time (not the precision of CGA file)? I have not found any function for this
Can you please try to repeat your test with my library libcgns_new.a statically linked or have you already done it?
Thanks in advance

**Comment by mennodeij** [comment]
It doesn't make sense I think to only use the libcgns-new.a file; without the header it is not complete and as you say the header contains the correct bitness.
Possibly this is a problem in cg_poly_elements_partial_read; between versions 3.x and 4.x the polyhedral layout has changed and the cgns library is backwards compatible. Maybe the backwards compatibility is faulty, I don't know...

**Comment by rsm73S** [comment]
I think I've found the reason. Might it be that the type of the last argument (parent_data) of cg_poly_elements_partial_read has been changed from int to long int? Anyway it seems to be inconsistent between cg_section_read and cg_poly_elements_partial_read. I used the same type (int) for both functions in my code. Now I must use int for cg_section_read and cgsize_t (long int ) for cg_poly_elements_partial_read.
Or may be cgsize_t was int before and has became at some moment long int.
But with correct type specification of parent_data arguments it works!
